ibis/expr/analysis.py

```diff
@@ -196,69 +196,30 @@ def apply_filter(op, predicates):
-    # change the shape and potential values of the data.
-    if any(
-        isinstance(
-            sel.arg if isinstance(sel, ops.Alias) else sel,
-            (ops.Unnest, ops.Window),
-        )
-        for sel in op.selections
-    ):
-        return default
-
-    # if any of the filter predicates have the parent expression among
-    # their roots, then pushdown (at least of that predicate) is not
-    # possible
-
-    # It's not unusual for the filter to reference the projection
-    # itself. If a predicate can be pushed down, in this case we must
-    # rewrite replacing the table refs with the roots internal to the
-    # projection we are referencing
-    #
-    # Assuming that the fields referenced by the filter predicate originate
-    # below the projection, we need to rewrite the predicate referencing
-    # the parent tables in the join being projected
-
-    # Potential fusion opportunity. The predicates may need to be
-    # rewritten in terms of the child table. This prevents the broken
-    # ref issue (described in more detail in #59)
-    try:
-        simplified_predicates = tuple(
-            sub_for(predicate, {op: op.table})
-            if not is_reduction(predicate)
-            else predicate
-            for predicate in predicates
-        )
-    except IntegrityError:
-        return default
-
-    if not shares_all_roots(simplified_predicates, op.table):
-        return default
-
-    # find spuriously simplified predicates
-    for predicate in simplified_predicates:
-        # find columns in the predicate
-        depends_on = predicate.find((ops.TableColumn, ops.Literal))
-        for projection in op.selections:
-            if not isinstance(projection, (ops.TableColumn, ops.Literal)):
-                # if the projection's table columns overlap with columns
-                # used in the predicate then we return immediately
-                #
-                # this means that we were too aggressive during simplification
-                # example: t.mutate(a=_.a + 1).filter(_.a > 1)
-                if projection.find((ops.TableColumn, ops.Literal)) & depends_on:
-                    return default
-
-    return ops.Selection(
-        op.table,
-        selections=op.selections,
-        predicates=op.predicates + simplified_predicates,
-        sort_keys=op.sort_keys,
-    )
+    projected_column_names = {
+        value.name
+        for value in parent.selections
+        if isinstance(value, ops.Value) and not isinstance(value, ops.TableColumn)
+    }
+    conflicting_projection = p.TableColumn(parent, In(projected_column_names))
+    pushdown_pattern = Eq(parent) >> parent.table
+
+    simplified = []
+    for pred in predicates:
+        if pred.match(
+            p.Reduction | p.Analytic | conflicting_projection, filter=p.Value
+        ):
+            return default
+
+        try:
+            simplified.append(pred.replace(pushdown_pattern))
+        except IntegrityError:
```

Comment:
I hate this, the `parent.table` is possibly a join, but occasionally the replacement works.